### PR TITLE
Use all gather in tensor metrics

### PR DESCRIPTION
**Pull request by samsja**
move from the all_reduce approach to a gather one where rank 0 gather every tensor over grad acc and gpu and then do distribution logging like max,min, mean but also median and quantile, can also do wandb histogram if we wanted to

### CODE REVIEW

File: src/prime_rl/trainer/loss.py [hunk 1]
```diff
@@ -89,13 +78,10 @@ def grpo_loss_ratio(
...
-    summed_loss = (loss * loss_mask).sum()
 
-    return summed_loss, {
-        "loss": loss.detach(),
-        "importance_ratio": importance_ratio.detach(),
-        "clipped_importance_ratio": clipped_importance_ratio.detach(),
-        "is_clipped": is_clipped.detach(),
+    return loss, {
+        "importance_ratio/importance_ratio": clipped_importance_ratio,
+        "importance_ratio/is_clipped": is_clipped,
```
Comment (mikasenghaas) [inline src/prime_rl/trainer/loss.py]: wait why are we not exporting the other tensors? now we are not logging `loss/{min,mean,max}`

Comment (mikasenghaas) [reply src/prime_rl/trainer/loss.py]: also i think we should embrace the sections and have each tensor that we log have its own section otherwise it will get messy soon with the amount of point metrics that we are logging now, so no nesting like `importance_ratio/...` imo

File: src/prime_rl/trainer/loss.py [hunk 2]
```diff
@@ -138,7 +124,6 @@ def selective_log_softmax(logits: Float[Tensor, "B L V"], index: Int[Tensor, "B
 
 
 @jaxtyped(typechecker=typechecker)
-@torch.no_grad()
```
Comment (mikasenghaas) [inline src/prime_rl/trainer/loss.py:141]: oh i liked this, looks much cleaner in trainer code. any reason for this?

File: src/prime_rl/trainer/train.py [hunk 3]
```diff
@@ -189,7 +188,7 @@ def train(config: TrainerConfig):
         # Optionally, compute the logprobs for the training batch
         compute_logprobs_time = 0
         num_micro_batches = len(micro_batches)
-        recomputed_logprob_errors = [torch.zeros_like(mb["logprobs"], device="cuda") for mb in micro_batches]
+        recomputed_logprob_errors = [torch.ones_like(mb["logprobs"], device="cuda") for mb in micro_batches]
```
Comment (mikasenghaas) [inline src/prime_rl/trainer/train.py:191]: i actually somewhat liked that the default is zero simply to show that we did not recompute. but yea maybe that's the more sane default given that we are logging a ratio

File: src/prime_rl/trainer/loss.py [hunk 4]
```diff
@@ -12,15 +12,13 @@ def compute_loss(
     logprobs: Float[Tensor, "B L"],
     old_logprobs: Float[Tensor, "B L"],
     advantages: Float[Tensor, "B L"],
-    loss_mask: Int[Tensor, "B L"],
     loss_config: LossConfig,
 ) -> tuple[Tensor, dict[str, Tensor]]:
```
Comment (mikasenghaas) [inline src/prime_rl/trainer/loss.py]: let's jaxtype the return type as well now that it is a B L tensor?

File: src/prime_rl/trainer/train.py [hunk 5]
```diff
@@ -260,65 +259,33 @@ def train(config: TrainerConfig):
...
+            loss_tensors["importance_ratio/recomputed_logprob_error"] = recomputed_logprob_errors[micro_step]
+            loss_tensors["importance_ratio/old_proba"] = torch.exp(old_logprobs)
+            loss_tensors["importance_ratio/proba"] = torch.exp(logprobs)
+            loss_tensors["entropy"] = entropy
+            loss_tensors["loss"] = per_token_loss
 
-            if micro_step == 0:
-                logger.debug("Logging loss tensors at first micro step")
-                for k, v in loss_tensors.items():
-                    logger.debug(f"\t{k}={lt.lovely(v)}")
```
Comment (mikasenghaas) [inline src/prime_rl/trainer/train.py:317]: RIP

Comment (mikasenghaas) [reply src/prime_rl/trainer/train.py:317]: why don't you like this? im thinking it might be nice to log the micro statistics to console in debug mode at least, o/w we have literlaly no way of knowing what is going on in the micro steps. will not be shown by default in info log mode ofc

File: src/prime_rl/trainer/utils.py [hunk 6]
```diff
@@ -114,64 +115,75 @@ def print_benchmark(history: dict[str, list[Any]]) -> None:
...
+            metrics[f"{k}/std"] = tensor.std().item()
+            metrics[f"{k}/min"] = tensor.min().item()
+            metrics[f"{k}/max"] = tensor.max().item()
+            metrics[f"{k}/median"] = torch.median(tensor).item()
+            metrics[f"{k}/1%"] = torch.quantile(tensor, 0.01).item()
+            metrics[f"{k}/5%"] = torch.quantile(tensor, 0.05).item()
+            metrics[f"{k}/10%"] = torch.quantile(tensor, 0.1).item()
+            metrics[f"{k}/90%"] = torch.quantile(tensor, 0.9).item()
+            metrics[f"{k}/95%"] = torch.quantile(tensor, 0.95).item()
+            metrics[f"{k}/99%"] = torch.quantile(tensor, 0.99).item()
```
Comment (mikasenghaas) [inline src/prime_rl/trainer/utils.py]: this seems like a LOT. im all for observability but having 11 plots for every tensor that we log (especially given that they all go in the same section atm) seems too much. also we seem to have lost the ability to specify which reductions we are interested, but ig we can also make the logging more verbose

File: src/prime_rl/trainer/loss.py (hunk 2 above)
Comment (samsja) [reply src/prime_rl/trainer/loss.py:141]: idk why it was here tbh

File: src/prime_rl/trainer/train.py (hunk 3 above)
Comment (samsja) [reply src/prime_rl/trainer/train.py:191]: hmmmm idk because the perfect score is 1 so would makes sense that we use 1 here

File: src/prime_rl/trainer/train.py (hunk 5 above)
Comment (samsja) [reply src/prime_rl/trainer/train.py:317]: feel free to add it back, was just me playing around and it was hard to read haha

File: src/prime_rl/trainer/utils.py (hunk 6 above)
Comment (samsja) [reply src/prime_rl/trainer/utils.py]: Its defo a lot but also the more concise, tbh max and min are useless most of the time. 

I think we should find a way to have a junk section where we put everything that we don't want to immediatily look at
